chia/wallet/util/puzzle_compression.py

```diff
@@ -0,0 +1,203 @@
+        else:
+            assert all(arg in solution_dict for arg in ["delegated_puzzle", "solution"])
+            return standard_puzzle.solution_for_delegated_puzzle(
+                Program.from_bytes(hexstr_to_bytes(solution_dict["delegated_puzzle"])),
+                Program.from_bytes(hexstr_to_bytes(solution_dict["solution"])),
+            )
+
+
+class CATPuzzle:
+    @staticmethod
+    def match(
+        puzzle: Program, compressor: "PuzzleCompressor"
+    ) -> Tuple[bool, List[Union[PuzzleRepresentation, Program]]]:
+        uncurried_mod, curried_args = puzzle.uncurry()
+        if CC_MOD == uncurried_mod:
+            tail_hash = curried_args.rest().first()
+            innerpuz = curried_args.rest().rest().first()
+            _, matched_inner = compressor.match_puzzle(innerpuz)
+            return True, [tail_hash, matched_inner]
+        else:
+            return False, []
+
+    @staticmethod
+    def construct(driver_dict, args: List[Union[PuzzleRepresentation, Program]]) -> Program:
+        assert isinstance(args[0], Program)
+        innerpuz = args[1]
+        if isinstance(args[1], PuzzleRepresentation):
+            innerpuz = args[1].construct(driver_dict)
+        return CC_MOD.curry(CC_MOD.get_tree_hash(), args[0].as_python(), innerpuz)
+
+    @staticmethod
+    def solve(driver_dict, args: List[Union[PuzzleRepresentation, Program]], solution_dict: Dict[str, str]) -> Program:
+        # TODO: implement this
+        return Program.to([])
+
+
+class OfferPuzzle:
+    @staticmethod
+    def match(
+        puzzle: Program, compressor: "PuzzleCompressor"
+    ) -> Tuple[bool, List[Union[PuzzleRepresentation, Program]]]:
+        if OFFER_MOD == puzzle:
+            return True, []
+        else:
+            return False, []
+
+    @staticmethod
+    def construct(driver_dict, args: List[Union[PuzzleRepresentation, Program]]) -> Program:
+        return OFFER_MOD
+
+    @staticmethod
+    def solve(driver_dict, args: List[Union[PuzzleRepresentation, Program]], solution_dict: Dict[str, str]) -> Program:
+        # TODO: implement this
+        return Program.to([])
+
+
+"""
+Below is a dictionary that contains version numbers mapped to dictionaries of identifiers mapped to puzzle drivers.
+The idea is that specifying a compression version will determine what puzzles you can interpret.
+All version numbers will support all versions lower than themselves to ensure backwards compatibility.
+It's a dict with numbers as the keys rather than a list to be clear that the order needs to be preserved.
+"""
+
+HASH_TO_DRIVER: Dict[uint16, Dict[bytes32, Any]] = {
+    uint16(0): {
+        standard_puzzle.MOD.get_tree_hash(): StandardPuzzle,
+        CC_MOD.get_tree_hash(): CATPuzzle,
+        OFFER_MOD.get_tree_hash(): OfferPuzzle,
+    }
+    # TODO: Add version tests when there are more versions :)
+}
+
+LATEST_VERSION: uint16 = uint16(max(HASH_TO_DRIVER.keys()))
+
+
+class CompressionVersionError(Exception):
+    pass
```

Comment:
Provide a `.__init__()` that takes the found version or somesuch and provides a useful string saying what version was found and what is supported.